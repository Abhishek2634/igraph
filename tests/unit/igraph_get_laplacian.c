/* -*- mode: C -*-  */
/*
   IGraph library.
   Copyright (C) 2022  The igraph development team <igraph@igraph.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

#include <igraph.h>
#include "test_utilities.h"

void test_laplacian(igraph_t *g, const igraph_vector_t *w, igraph_bool_t dir, igraph_laplacian_normalization_t normalization) {
    igraph_matrix_t m;

    igraph_matrix_init(&m, 0, 0);

    igraph_get_laplacian(g, &m, IGRAPH_OUT, normalization, w);
    igraph_matrix_print(&m);

    igraph_matrix_destroy(&m);
}

int main() {
    igraph_t g_un, g_dir;
    igraph_vector_t weights;

    igraph_vector_init_int(&weights, 9, 1, 2, 3, 4, 5, 2, 2, 3, 3);
    igraph_small(&g_un, 6, IGRAPH_UNDIRECTED, 0,1, 1,2, 2,3, 3,4, 4,0, 1,1, 2,2, 1,2, 3,4, -1);
    igraph_small(&g_dir, 6, IGRAPH_DIRECTED, 0,1, 1,2, 2,3, 3,4, 4,0, 1,1, 2,2, 1,2, 3,4, -1);

    for (int i = 0; i < 8; i++) {
        igraph_bool_t is_normalized = i / 4;
        igraph_vector_t* v = ((i & 2) / 2 ? &weights : 0);
        igraph_bool_t dir = (i % 2 ? IGRAPH_DIRECTED : IGRAPH_UNDIRECTED);

        printf("=== %sormalized, %sweighted, %sdirected\n",
               (is_normalized ? "N" : "Unn"),
               (v != 0 ? "" : "un"),
               (dir == IGRAPH_DIRECTED ? "" : "un")
              );

        test_laplacian(dir ? &g_dir : &g_un, v, dir, is_normalized ? IGRAPH_LAPLACIAN_SYMMETRIC : IGRAPH_LAPLACIAN_UNNORMALIZED);
    }

    igraph_vector_destroy(&weights);
    igraph_destroy(&g_un);
    igraph_destroy(&g_dir);

    VERIFY_FINALLY_STACK();

    return 0;
}
