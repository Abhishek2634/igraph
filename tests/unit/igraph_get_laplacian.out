=== Unnormalized, unweighted, undirected
2 -1 0 0 -1 0
-1 3 -2 0 0 0
0 -2 3 -1 0 0
0 0 -1 3 -2 0
-1 0 0 -2 3 0
0 0 0 0 0 0
=== Unnormalized, unweighted, directed
1 -1 0 0 0 0
0 2 -2 0 0 0
0 0 1 -1 0 0
0 0 0 2 -2 0
-1 0 0 0 1 0
0 0 0 0 0 0
=== Unnormalized, weighted, undirected
6 -1 0 0 -5 0
-1 6 -5 0 0 0
0 -5 8 -3 0 0
0 0 -3 10 -7 0
-5 0 0 -7 12 0
0 0 0 0 0 0
=== Unnormalized, weighted, directed
1 -1 0 0 0 0
0 5 -5 0 0 0
0 0 3 -3 0 0
0 0 0 7 -7 0
-5 0 0 0 5 0
0 0 0 0 0 0
=== Normalized, unweighted, undirected
1 -0.316228 0 0 -0.408248 0
-0.316228 0.6 -0.4 0 0 0
0 -0.4 0.6 -0.258199 0 0
0 0 -0.258199 1 -0.666667 0
-0.408248 0 0 -0.666667 1 0
0 0 0 0 0 0
=== Normalized, unweighted, directed
1 -0.57735 0 0 0 0
0 0.666667 -0.816497 0 0 0
0 0 0.5 -0.5 0 0
0 0 0 1 -1.41421 0
-1 0 0 0 1 0
0 0 0 0 0 0
=== Normalized, weighted, undirected
1 -0.129099 0 0 -0.589256 0
-0.129099 0.6 -0.456435 0 0 0
0 -0.456435 0.666667 -0.273861 0 0
0 0 -0.273861 1 -0.63901 0
-0.589256 0 0 -0.63901 1 0
0 0 0 0 0 0
=== Normalized, weighted, directed
1 -0.377964 0 0 0 0
0 0.714286 -0.845154 0 0 0
0 0 0.6 -0.507093 0 0
0 0 0 1 -1.18322 0
-2.23607 0 0 0 1 0
0 0 0 0 0 0
